pulser-simulation/pulser_simulation/simulation.py

```diff
@@ -492,19 +492,35 @@ def run(
 
                 .. _docs: https://bit.ly/3il9A2u
         """
-        if "max_step" not in options:
-            pulse_durations = [
-                slot.tf - slot.ti
-                for ch_sample in self.samples_obj.samples_list
-                for slot in ch_sample.slots
-                if not (
-                    np.all(np.isclose(ch_sample.amp[slot.ti : slot.tf], 0))
-                    and np.all(np.isclose(ch_sample.det[slot.ti : slot.tf], 0))
+
+        def get_min_variation(ch_sample: ChannelSamples) -> int:
+            end_point = ch_sample.duration - 1
+            min_variations: list[int] = []
+            for sample in (ch_sample.amp, ch_sample.det):
+                # Find the indices at which the sample varies
+                variations_indices: np.ndarray = np.concatenate(
+                    [
+                        np.array([-1]),
+                        np.nonzero(np.diff(sample)),
+                        np.array([end_point]),
+                    ],
+                    axis=None,
```

Comment:
I think you can use the `prepend` and `append` arguments in the `np.diff()` down below to avoid this concatenation

Reply:
I think so too ;)